eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import numpy as np
+import keras
+import keras.backend as K
+from keras.models import Model
+from keras.layers import Layer, Lambda
+from keras.preprocessing.image import load_img, img_to_array, array_to_img
+
+from eli5.base import Explanation
+from eli5.explain import explain_prediction
+
+
+DESCRIPTION_KERAS = """Grad-CAM visualization for image input; output is images"""
+
+# note that keras.models.Sequential subclasses keras.models.Model, so we can just register Model
+@explain_prediction.register(Model)
+def explain_prediction_keras(estimator, doc,
+                             target_names=None, # TODO: implement this
+                             targets=None,
+                             # new parameters:
+                             layer=None,
+                            ):
+    """
+    Explain an image prediction of a Keras image classifier.
+
+    See :func:`eli5.explain_prediction` for more information about the ``estimator``,
+    ``doc``, ``target_names``, and ``targets`` parameters.
+
+    Parameters
+    ----------
+    estimator : object
+        Instance of a Keras neural network model.
+    doc : object
+        An input image as a tensor to ``estimator``, for example ``numpy.ndarray``.
+    target_names : list, optional
+        *`Not Implemented`*.
+
+        Names for classes in the final output layer.
+    targets : list[int], optional
+        Prediction ID's to focus on.
+
+        *`Currently only the first prediction from the list is explained`*.
+
+        If None, the model is fed the input and its top prediction 
+        is taken as target as the target automatically.
+    layer : int or str or object, optional
+        The activation layer in the model to perform Grad-CAM on,
+        a valid keras layer name, layer index, or an instance of keras.layers.Layer.
+        
+        If None, a suitable layer is attempted to be retrieved (raise ValueError if can not).
+
+    Returns
+    -------
+    A :class:`base.Explanation` object with the ``image`` and ``heatmap`` attributes set.
+    """
+    activation_layer = get_activation_layer(estimator, layer)
+    predicted = get_target_prediction(estimator, doc, targets)
+    
+    heatmap = grad_cam(estimator, doc, predicted, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations' (the output is not yet a heat map)
+
+    # need to insert a 'channel' axis to get a rank 3 image from rank 2
+    heatmap = np.expand_dims(heatmap, axis=-1)
+    heatmap = array_to_img(heatmap) # -> PIL image
+    
+    # take the single image from the input 'batch' (rank 4 to rank 3)
+    doc = doc[0]
```

Comment:
I think it would be nice to clarify that a batch of size one needs to be passed in the API docs, and also to check that the user did pass a batch of size one, raising an error if not. I'm not sure if some models could work with an image without batch dimension - if yes, we might want to check the shape here.

Reply:
Good catch. We could rely on Keras exceptions. For example, this seems pretty self-explanatory:
![Screenshot from 2019-06-13 16-07-35](https://user-images.githubusercontent.com/39887323/59444517-ed584600-8df5-11e9-8b1b-752e4973de5c.png)

However, it can be esoteric too:
![Screenshot from 2019-06-13 16-08-05](https://user-images.githubusercontent.com/39887323/59444518-ed584600-8df5-11e9-9729-dfe7801cf0b3.png)

So I will add our own validation function.

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/2ce3ac396cbe49202eb73b18639f2aaa815baa29